File: services/data-pipeline-py/app/schemas/strategy.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class StrategySignal(BaseModel):
    model_config = ConfigDict(extra="forbid")

    strategy_id: str
    signal_id: str
    symbol: str
    side: str
    quantity: float = 0.0
    notional: float = 0.0
    price_hint: float = 0.0
    timestamp: datetime
    scenario: str = "baseline"


class PublishSignalsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scenario: str = "baseline"
    strategy_id: str | None = None
    limit: int = 0
    offset: int = 0


class ReplaySignalsRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    scenario: str = "baseline"
    strategy_id: str | None = None
    speed_multiplier: float = Field(default=0.0, ge=0.0, le=1000.0)

File: services/simulator-ui/src/shared/types/strategy.ts
import type { IsoDateTimeString, QueryParams } from "./api";
import type { OrderSide } from "./paper";

export interface StrategySignal {
  strategy_id: string;
  signal_id: string;
  symbol: string;
  side: OrderSide;
  quantity: number;
  notional: number;
  price_hint: number;
  timestamp: IsoDateTimeString;
  scenario?: string;
}

export interface StrategySignalsQuery extends QueryParams {
  scenario?: string;
  strategy_id?: string | null;
  limit?: number;
  offset?: number;
}

export interface PublishSignalsRequest {
  scenario?: string;
  strategy_id?: string | null;
  limit?: number;
  offset?: number;
}

export interface ReplaySignalsRequest {
  scenario?: string;
  strategy_id?: string | null;
  speed_multiplier: number;
}

export interface StrategySignalsResponse {
  signals: StrategySignal[];
}

export interface StrategyScenariosResponse {
  scenarios: string[];
}

export interface PublishSignalsResponse {
  scenario: string;
  published_count: number;
  signal_ids: string[];
  strategy_ids: string[];
  transports: string[];
}

export interface ReplaySignalsResponse {
  scenario: string;
  published_count: number;
  signal_ids: string[];
  strategy_id?: string | null;
  speed_multiplier: number;
}
